### DISCUSSION TITLE
Modify mass matrix to model reflected inertia

### DISCUSSION THREAD

**Discussion by ZhaomingXie** [bulletphysics/bullet3#3615] [General]
Hi
We are trying to build a pybullet model for the cassie robot, and we believe reflected inertia is an important parameter. I wonder if there is a way to model reflected inertia in bullet? 
My understanding is that it is equivalent to adding the reflected inertia value to the corresponding diagonal element in the mass matrix, as mentioned in equation 2 in this paper: https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=7140004. I believe this is equivalent to the armature parameter in Mujoco, as discussed in this post by Emo:  http://www.mujoco.org/forum/index.php?threads/armature.3378/. 
I looked through the API for pybullet and didn't find an equivalence. I wonder if there is an easy way to implement this functionality.

**Comment by erwincoumans** [comment]
At the moment, you need to model the rotor explicitly and add a gearing constraint to model the gearing ratio. This will have the effect of reflected inertia. See also the description in the Modern Robotics book (http://modernrobotics.org, there is a free pdf of the book)
We can also implement the gearing ratio directly into the Articulated Body Algorithm, as described in section 8.9.2 of that book. This is done in the MIT Mini Cheetah simulator.
See the gearing ratio in this implementation:
https://github.com/mit-biomimetics/Cheetah-Software/blob/master/common/src/Dynamics/FloatingBaseModel.cpp#L479

I'll look into having this also in PyBullet. Until then, try modeling the rotor and add a gearing constraint.
See the [mimicJointConstraint.py](https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/examples/mimicJointConstraint.py) for an example, the constraint supports a gearing ratio.

**Reply by eshafeeqe** [reply]
Hello, @erwincoumans, 

I also came across the same issue, I am assuming that this functionality has not been implemented in Pybullet yet. Can you give an example of how to model rotor inertia and gear ratio using the existing Pybullet API?
